### DISCUSSION TITLE
Accesso automatico senza consenso

### DISCUSSION THREAD

**Discussion by elenalic** [fattureincloud/api#283] [Q&A]
Buongiorno,

Ho notato che quando voglio autorizzare il collegamento e acconsentire al trattamento dei dati da parte dell'azienda, in automatico si connette all'azienda  senza che io abbia cliccato sul button "Autorizza" oppure senza che il flag "consenti automaticamente" sia attivo.

Come è possibile questo comportamento? 

Grazie,
Elena

**Comment by elenalic** [comment]
Volevo aggiungere che ho provato anche a fare svariati accessi in incognito, ma il comportamento è sempre il medesimo.

Grazie

**Comment by elenalic** [comment]
![image](https://user-images.githubusercontent.com/126566722/221843590-4f056ef3-5189-4470-a742-646241d8f515.png)

**Comment by valmoz** [comment]
Buongiorno @elenalic 
chiedo scusa, ma ciò che dice non mi torna.

Ho appena testato creando una nuova applicazione con OAuth abilitato e aprendo il link di redirect come segue (ho cambiato il client id):
```
https://api-v2.fattureincloud.it/oauth/authorize?response_type=code&client_id=FAKE_ID&redirect_uri=https%3A%2F%2Fwww.google.com&scope=situation%3Ar%20entity.clients%3Aa%20issued_documents.quotes%3Aa&state=EXAMPLE_STATE
```
La pagina si è caricata correttamente, come previsto si è fermato sulla pagina di richiesta autorizzazione (come da suo screenshot).
Ho provato poi a chiudere la finestra oppure cliccare su "annulla", e come pensavo non è stato ritornato il code necessario ad ottenere l'access token, e l'applicazione non appare nell'elenco delle applicazioni collegate.

Potrebbe indicarmi le operazioni effettuate per riprodurre il problema?

Si assicuri inoltre di non aver mai associato l'account Fatture in Cloud che sta utilizzando con la sua applicazione, il comportamento potrebbe essere legato al fatto che lei ha già assegnato i permessi alla applicazione in uso (può controllare tramite la sezione "applicazioni collegate").

Resto in attesa di un riscontro.

Grazie
Mauro

**Comment by elenalic** [comment]
Buon pomeriggio @valmoz ,

Allora le operazioni che devo effettuare per potermi connettere sono le seguenti:
1. passo nell'API reference (https://developers.fattureincloud.it/api-reference/#overview) client ID e client Secret della mia applicazione
2. Metto le mie credenziali nella schermata di Log in di Fatture in Cloud e avvio il processo di collegamento
3. si apre, velocissimamente, la pagine che ho mostrato nell'immagine in precedenze e poi si connette in modo automatico senza che io possa cliccare in precedenza su Autorizza oppure flaggare il check box
4. il collegamento è avvenuto correttamente, ma senza autorizzazione

Resto in attesa di un riscontro,
grazie molto 

Elena

**Reply by valmoz** [reply]
Allora, come già detto in precedenza, nella sezione "Applicazioni collegate" è possibile vedere che l'applicazione che sta utilizzando è già collegata con il suo account.

Quando l'applicazione è già collegata possiede già le autorizzazioni necessarie, per questo motivo la pagina appare ma subito dopo viene effettuato il redirect e quindi da l'impressione che l'autorizzazione viene saltata.

Le consiglio di provare a scollegare l'applicazione dal suo account Fatture in Cloud dalla sezione "Applicazioni Collegate" e di riprovare ad effettuare la connessione, la pagina questa volta dovrebbe richiedere di premere il pulsante.

Le consiglio anche di provare a utilizzare un altro account di test per verificare che tutto funzioni come si deve.

Grazie
Mauro

**Reply by valmoz** [reply]
Buongiorno @elenalic 
abbiamo chiarito i suoi dubbi?

Grazie
Mauro

**Comment by elenalic** [comment]
Buongiorno @valmoz ,

Ho rifatto altri test. 

Può essere che accettando i cookies si ricordi del mio precedente login e per cui la mia precedente autorizzazione. 

Ma dato che con API v.2 offre la possibilità di connettere diverse aziende da uno stesso account, non trovo corretto il comportamento che "ricordandosi" non mi fa più autorizzare, anche perchè dato il cambio di azienda devo aver la possibilità di dare l'autorizzazione. Inoltre non è spuntato il flag riguardo "concedi automaticamente".

Grazie, 
Elena

**Reply by emajo** [reply]
Ciao @elenalic, no, non si tratta di una questione di cookies, semplicemente se l'applicazione è già stata collegata non verrà più proposta la schermata di autorizzazione, per togliere l'autorizzazione a una azienda o dare accesso ad un'altra ciò risulta possibile dalle impostazione di fatture in cloud nella sezione applicazioni collegate.

Grazie,
Emanuele.

**Reply by elenalic** [reply]
Buongiorno @emajo,

io continuo a testare ma il comportamento rimane il medesimo. il pop up si apre, non mi fa selezionare l'azienda con la quale vorrei connettermi e si connette in automatico.
Anche se la volta precedente aveva aspettato un mio consenso. 

Diciamo che è un comportamento discontinuo e randomico, ma ripeto non è una mia "impressione". 

Quindi se non è un discorso di cookies che cosa potrebbe essere?

Grazie,
Elena.

**Reply by emajo** [reply]
Buongiorno @elenalic, come già spiegato nel messaggio precedente quel 'popup' viene mostrato solamente la prima volta (ovvero quando l'applicazione non è già connessa all'account), successivamente come riscontrato da lei viene aperto e chiuso connettendosi in automatico (appunto perché l'applicazione risulta già collegata), può testare questo comportamento andando a disconnettere manualmente dalle impostazioni l'applicazione in questione, vedrà che la volta successiva le verrà chiesta l'autorizzazione di connessione.

Per andare a cambiare le aziende su cui è connessa l'applicazione è necessario farlo dalle impostazioni, dove nella sezione applicazioni collegate è possibile fare questa operazione.

Questo comportamento è voluto da noi, non si tratta di un bug, posso chiederle per quale motivo le risulta necessario cambiare company ogni volta? Questo meccanismo dovrebbe essere gestito direttamente dall'integrazione (recuperando la lista delle companies per cui è stata data l'autorizzazione e chiedendo all'utente quale vuole utilizzare) non andando a cambiare la company a cui è collegata l'integrazione.

Grazie,
Emanuele.

**Comment by elenalic** [comment]
<img width="847" alt="image" src="https://user-images.githubusercontent.com/126566722/224307329-14456b1f-8812-4db1-a9bc-4fdb41997457.png">
<img width="745" alt="1489_828_1 5" src="https://user-images.githubusercontent.com/126566722/224307649-7e461c19-fbd4-4b69-a06c-3005e88b1ed5.png">

Ciao @emajo, 
ti ho allegato le due foto per dimostrarti che nonostante io abbia scollegato l'app come da te suggerito, il comportamento rimane il medesimo. Per il resto mi risulta importante il cambio di company per poterlo gestire questa vostra novità, cioè l'inserimento di più company.

Grazie,
Elena

**Reply by emajo** [reply]
Mi scusi ma dagli screenshots allegati mi sembra che il comportamento sia proprio quello atteso, dopo aver scollegato l'applicazione viene chiesta l'autorizzazione per il nuovo collegamento.

L'inserimento di più companies non va gestito tramite il collegamento dell'app, il flusso corretto è il seguente:

1. L'utente collega l'applicazione la prima volta, gli viene chiesto quali companies vuole collegare da quel popup e ne seleziona 3
2. L'integrazione fa una chiamata di list user companies, questa chiamata ritornerà appunto le 3 companies per cui l'utente ha dato l'autorizzazione
3. L'integrazione fa selezionare all'utente quale delle 3 companies recuperate vuole usare, e viene usato il relativo company id per effettuare le chiamate successive
4. Se l'utente vuole cambiare company, non deve rieffettuare il flusso di collegamento, è l'integrazione che deve offrigli la possibilità di cambiare company sul quale vuole che operi, dato che l'autorizzazione è già stata fornita

da quello che ho capito io il flusso che attualmente vuole implementare è:

1. L'utente collega l'applicazione la prima volta e seleziona una company
2. L'integrazione opera esclusivamente per quella company
3. Se l'utente deve cambiare company su cui operare deve rifare il flusso di connessione scegliendo un altra company

se è così, non è il modo corretto di implementare questa funzionalità, la schermata di connessione dell'applicazione l'utente la vedrà effettivamente solo una volta (alla prima connessione dell'applicazione), in caso avesse altri dubbi le consiglio di scrivere al nostro supporto clienti contattandoli tramite la chat che può trovare sul nostro sito.

Grazie,
Emanuele.

**Reply by elenalic** [reply]
No ho fatto lo screen molto velocemente per mostrare che non c'è il flag sul concedere in maniera automatica, ma mi appare il pop up e poi scompare senza che io possa fare niente.

farò dei test date le tue informazioni.

Grazie,
Elena

**Comment by Zanfo96** [comment]
Hi @emajo, 

Let's suppose we have the following scenario:

My FiC user has 3 companies. I created an application for the first company and integrated that with my backend. I proceed to connect it to that user, and in authorization page after login I have this flag "concedi automaticamente l'accesso a nuove aziende" and i proceed to flag it. This to my understanding means the following:

if some day later I create a new company in that user, I won't have to authorize it never again and that's fine. But the first time i have the "parent flag" selected, every company in my user will be forever authorized, just because my application belongs to that user and i did not open the dropdown. To me it's working, but it's not that clear: if i'm an accountant and want to authorize just a company and I do not proceed to deselect the other companies, I will forever be authorized with other companies unless I take actions in FiC account.

Do you maybe have a specific documentation for this authorization process?

Thanks and have a good weekend, 

Denis
